llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -2147,14 +2148,45 @@ static SDValue foldAddSubBoolOfMaskedVal(SDNode *N, SelectionDAG &DAG) {
   return DAG.getNode(IsAdd ? ISD::SUB : ISD::ADD, DL, VT, C1, LowBit);
 }
 
+/// Try to fold a pointer arithmetic node, preferring integer arithmetic.
+/// This needs to be done separately from normal addition, because pointer
+/// addition is not commutative.
 SDValue DAGCombiner::visitPTRADD(SDNode *N) {
   SDValue N0 = N->getOperand(0);
   SDValue N1 = N->getOperand(1);
+  EVT PtrVT = N0.getValueType();
+  EVT IntVT = N1.getValueType();
+  SDLoc DL(N);
+
+  // fold (ptradd undef, y) -> undef
+  if (N0.isUndef())
+    return N0;
+
+  // fold (ptradd x, undef) -> undef
+  if (N1.isUndef())
+    return DAG.getUNDEF(PtrVT);
+
   // fold (ptradd x, 0) -> 0
-  // This needs to be done separately from normal addition, because pointer
-  // addition is not commutative.
   if (isNullConstant(N1))
     return N0;
+
+  // Reassociate: (ptradd (ptradd x, y), z) -> (ptradd x, (add y, z)) if:
+  //   * x is a null pointer; or
+  //   * y and z are constants; or
+  //   * y is a constant and (ptradd x, y) has one use.
```

Comment:
Does it makes sense to also do it if both y and z have only one use?

Reply:
My intuition says no, as it forces `y` to be live for longer, now alongside `(ptradd x, y)`, so increases register pressure. A case we could do is `(ptradd x, y) and z have only one use`, though. I also realise I didn't quite implement what the comment says; I'm doing `y is a constant and z has one use` not `y is a constant and (ptradd x, y) has one use`, though both make sense to do.